src/adc.rs

```diff
@@ -241,6 +254,52 @@ impl VTemp {
     pub fn disable(&mut self, adc: &mut Adc) {
         adc.rb.ccr.modify(|_, w| w.tsen().clear_bit());
     }
+
+    pub fn is_enabled(&self, adc: &Adc) -> bool {
+        adc.rb.ccr.read().tsen().bit_is_set()
+    }
+
+    pub fn convert_temp(vtemp: u16, vdda: u16) -> i16 {
+        const VDD_CALIB: i32 = 3300_i32;
+        let vtemp30_cal = i32::from(unsafe { ptr::read(VTEMPCAL30) }) * 100;
+        let vtemp110_cal = i32::from(unsafe { ptr::read(VTEMPCAL110) }) * 100;
+
+        let mut temperature: i32 = (vtemp as i32) * 100;
+        temperature = (temperature * (vdda as i32) / VDD_CALIB) - vtemp30_cal;
+        temperature *= (110 - 30) * 100;
+        temperature /= vtemp110_cal - vtemp30_cal;
+        temperature += 3000;
+        temperature as i16
+    }
+
+    pub fn read_vtemp(adc: &mut Adc) -> i16 {
+        let mut vtemp = Self::new();
+        let vtemp_preenable = vtemp.is_enabled(&adc);
+
+        if !vtemp_preenable {
+            vtemp.enable(adc);
+
+            // Double read of vdda to allow sufficient startup time for the temp sensor
```

Comment:
Wouldn't using a delay reference be nicer here?

Reply:
Yes, though I was trying to avoid forcing the user to provide a delay if possible.

Reply:
Made an option